File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/util/LabelWrapper.java
/*
 * Copyright (c) 2005, The JUNG Authors
 * All rights reserved.
 *
 * This software is open-source under the BSD license; see either "license.txt"
 * or https://github.com/jrtom/jung/blob/master/LICENSE for a description.
 *
 */
package edu.uci.ics.jung.visualization.util;

import com.google.common.base.Function;

/**
 * A utility to wrap long lines, creating html strings
 * with line breaks at a settable max line length
 * 
 * @author Tom Nelson - tomnelson@dev.java.net
 *
 */
public class LabelWrapper implements Function<String,String> {

	int lineLength;
	public static final String breaker = "<p>";
	
	/**
	 * Create an instance with default line break length = 10
	 *
	 */
	public LabelWrapper() {
		this(10);
	}
	
	/**
	 * Create an instance with passed line break length
	 * @param lineLength the max length for lines
	 */
	public LabelWrapper(int lineLength) {
		this.lineLength = lineLength;
	}

	/**
	 * call 'wrap' to transform the passed String
	 */
	public String apply(String str) {
		if(str != null) {
			return wrap(str);
		} else {
			return null;
		}
	}
	
	/**
	 * line-wrap the passed String as an html string with
	 * break Strings inserted.
	 * 
	 * @param str
	 * @return
	 */
	private String wrap(String str) {
		StringBuilder buf = new StringBuilder(str);
		int len = lineLength;
		while(len < buf.length()) {
			int idx = buf.lastIndexOf(" ", len);
			if(idx != -1) {
				buf.replace(idx, idx+1, breaker);
				len = idx + breaker.length() +lineLength;
			} else {
				buf.insert(len, breaker);
				len += breaker.length() + lineLength;
			}
		}
		buf.insert(0, "<html>");
		return buf.toString();
	}
	
	public static void main(String[] args) {
		String[] lines = {
				"This is a line with many short words that I will break into shorter lines.",
				"thisisalinewithnobreakssowhoknowswhereitwillwrap",
				"short line"
		};
		LabelWrapper w = new LabelWrapper(10);
		for(int i=0; i<lines.length; i++) {
			System.err.println("from "+lines[i]+" to "+w.wrap(lines[i]));
		}
	}
}

File: jung-visualization/src/main/kotlin/edu/uci/ics/jung/visualization/util/LabelWrapper.kt
/*
 * Copyright (c) 2005, The JUNG Authors
 * All rights reserved.
 *
 * This software is open-source under the BSD license; see either "license.txt"
 * or https://github.com/jrtom/jung/blob/master/LICENSE for a description.
 */
package edu.uci.ics.jung.visualization.util

import java.util.function.Function

/**
 * A utility to wrap long lines, creating html strings with line breaks at a settable max line
 * length.
 *
 * @author Tom Nelson - tomnelson@dev.java.net
 */
open class LabelWrapper @JvmOverloads constructor(
    private val lineLength: Int = 10
) : Function<String?, String?> {

    /** Call 'wrap' to transform the passed String. */
    override fun apply(str: String?): String? = str?.let { wrap(it) }

    /**
     * Line-wrap the passed String as an html string with break Strings inserted.
     */
    private fun wrap(str: String): String {
        val buf = StringBuilder(str)
        var len = lineLength
        while (len < buf.length) {
            val idx = buf.lastIndexOf(" ", len)
            if (idx != -1) {
                buf.replace(idx, idx + 1, BREAKER)
                len = idx + BREAKER.length + lineLength
            } else {
                buf.insert(len, BREAKER)
                len += BREAKER.length + lineLength
            }
        }
        buf.insert(0, "<html>")
        return buf.toString()
    }

    companion object {
        @JvmField
        val BREAKER: String = "<p>"

        @JvmStatic
        fun main(args: Array<String>) {
            val lines = arrayOf(
                "This is a line with many short words that I will break into shorter lines.",
                "thisisalinewithnobreakssowhoknowswhereitwillwrap",
                "short line"
            )
            val w = LabelWrapper(10)
            for (line in lines) {
                System.err.println("from $line to ${w.wrap(line)}")
            }
        }
    }
}
